service/src/event/ordering_task.rs

```diff
@@ -175,302 +97,470 @@ impl OrderingTask {
+                    .cid_map
+                    .get(now_ready)
+                    .expect("must have value in cid_map if it's in prev_map")
+                    .to_owned();
+                match ev {
+                    StreamEvent::InitEvent(cid) => {
+                        unreachable!("should not have found an undelivered init event and added it to our delivery queue {}", cid);
+                    }
+                    StreamEvent::KnownDeliverable(_) => {
+                        // This is fine as we could have already discovered and added ourself to the queue above.
+                        // We get marked as KnownDeliverable in that case and we don't have anything more to do.
+                    }
+                    StreamEvent::Undelivered(meta) => {
+                        // Discovering this event's prev (and therefore this event) as deliverable might have unlocked
+                        // something else, so we add it to the back of the queue to check.
+                        newly_ready.push_back(meta.cid);
+                        self.discovered_deliverable_prev(meta.cid);
+                    }
+                }
+            }
         }
+        debug!(count=%self.new_deliverable.len(), "deliverable events discovered");
+        Ok(())
     }
 }
 
+#[derive(Debug)]
+pub struct OrderingState {
+    pending_by_stream: HashMap<StreamCid, StreamEvents>,
+    deliverable: VecDeque<EventCid>,
+}
+
 impl OrderingState {
-    pub fn new() -> Self {
+    fn new() -> Self {
         Self {
             pending_by_stream: HashMap::new(),
-            ready_events: VecDeque::new(),
+            deliverable: VecDeque::new(),
         }
     }
 
-    /// This will review all the events for any streams known to have undelivered events and see if any of them are now deliverable.
-    /// If `streams_to_process` is None, all streams will be processed, otherwise only the streams in the set will be processed.
-    /// Processing all streams could take a long time and not necessarily do anything productive (if we're missing a key event, we're still blocked).
-    /// However, passing a value for `streams_to_process` when we know something has changed is likely to have positive results and be much faster.
-    pub(crate) async fn process_events(
-        &mut self,
-        pool: &SqlitePool,
-        streams_to_process: Option<HashSet<InitCid>>,
-    ) -> Result<()> {
-        self.persist_ready_events(pool).await?;
-        for (cid, stream_events) in self.pending_by_stream.iter_mut() {
-            if streams_to_process
-                .as_ref()
-                .map_or(false, |to_do| !to_do.contains(cid))
-            {
-                continue;
-            }
-            let deliverable = Self::discover_deliverable_events(pool, stream_events).await?;
-            if !deliverable.is_empty() {
-                self.ready_events.extend(deliverable)
-            }
+    /// Add a stream to the list of streams to process.
```

Comment:
```suggestion
    /// Add streams to the list of streams to process.
```

Reply:
looks like the events passed here might not all belong to the same stream right?

Reply:
yeah, that's correct. modified to read

```
    /// Update the list of streams to process with new events.
    /// Relies on `add_stream_event` to handle updating the internal state.
```